l1-contracts/contracts/bridge/AssetRouterBase.sol

```diff
@@ -0,0 +1,179 @@
+    constructor(IBridgehub _bridgehub) {
+        BRIDGE_HUB = _bridgehub;
+    }
+
+    /// @notice Sets the L1ERC20Bridge contract address.
+    /// @dev Should be called only once by the owner.
+    /// @param _nativeTokenVault The address of the native token vault.
+    function setNativeTokenVault(IL1NativeTokenVault _nativeTokenVault) external onlyOwner {
+        require(address(nativeTokenVault) == address(0), "ShB: native token vault already set");
+        require(address(_nativeTokenVault) != address(0), "ShB: native token vault 0");
+        nativeTokenVault = _nativeTokenVault;
+    }
+
+    /// @notice Sets the asset handler address for a given asset ID.
+    /// @dev No access control on the caller, as msg.sender is encoded in the assetId.
+    /// @param _assetData In most cases this parameter is bytes32 encoded token address. However, it can include extra information used by custom asset handlers.
+    /// @param _assetHandlerAddress The address of the asset handler, which will hold the token of interest.
+    function setAssetHandlerAddress(bytes32 _assetData, address _assetHandlerAddress) external virtual {
+        address sender = msg.sender == address(nativeTokenVault) ? L2_NATIVE_TOKEN_VAULT_ADDRESS : msg.sender;
+        bytes32 assetId = keccak256(abi.encode(uint256(block.chainid), sender, _assetData));
+        assetHandlerAddress[assetId] = _assetHandlerAddress;
+        emit AssetHandlerRegistered(assetId, _assetHandlerAddress, _assetData, sender);
+    }
+
+    /// @notice Allows bridgehub to acquire mintValue for L1->L2 transactions.
+    /// @dev If the corresponding L2 transaction fails, refunds are issued to a refund recipient on L2.
+    /// @param _chainId The chain ID of the ZK chain to which deposit.
+    /// @param _assetId The deposited asset ID.
+    /// @param _prevMsgSender The `msg.sender` address from the external call that initiated current one.
+    /// @param _amount The total amount of tokens to be bridged.
+    function bridgehubDepositBaseToken(
+        uint256 _chainId,
+        bytes32 _assetId,
+        address _prevMsgSender,
+        uint256 _amount
+    ) external payable virtual;
+
+    /// @notice Initiates a deposit transaction within Bridgehub, used by `requestL2TransactionTwoBridges`.
+    /// @param _chainId The chain ID of the ZK chain to which deposit.
+    /// @param _prevMsgSender The `msg.sender` address from the external call that initiated current one.
+    /// @param _l2Value The L2 `msg.value` from the L1 -> L2 deposit transaction.
+    /// @param _data The calldata for the second bridge deposit.
+    /// @return request The data used by the bridgehub to create L2 transaction request to specific ZK chain.
+    function bridgehubDeposit(
+        uint256 _chainId,
+        address _prevMsgSender,
+        uint256 _l2Value,
+        bytes calldata _data
+    ) external payable virtual returns (L2TransactionRequestTwoBridgesInner memory request);
```

Comment:
This should not be virtual, is there a problem implementing it here?

Reply:
The issue is that bridgehubDeposit has some legacy decoding parsing. Why should all the asset routers inherit this functionality?

bridgehubDepositBaseToken though shouldn't be virtual

I perceived that L1AssetRouter would implement bridgehubDeposit with legacy vs L2 (or ZKAssetRouter) implements bridgehubDeposit without legacy code, wdyt?